python/pyarrow/array.pxi

```diff
@@ -824,6 +824,46 @@ cdef class Array(_PandasConvertible):
 
         return wrap_datum(out)
 
+    def filter(self, Array mask):
+        """
+        Filter the array with a boolean mask.
+
+        Parameters
+        ----------
+        mask : Array
+            The boolean mask indicating which values to extract.
+
+        Returns
+        -------
+        Array
+
+        Examples
+        --------
+
+        >>> import pyarrow as pa
+        >>> arr = pa.array(["a", "b", "c", None, "e"])
+        >>> mask = pa.array([True, False, None, False, True])
+        >>> arr.filter(mask)
+        <pyarrow.lib.StringArray object at 0x7fa826df9200>
+        [
+          "a",
+          null,
+          "e"
+        ]
+        """
+        cdef:
+            cdef CDatum out
+
+        if mask.type.id != _Type_BOOL:
+            raise TypeError(
+                "The mask should be boolean type, got {}".format(mask.type))
```

Comment:
I was wondering, is this a check that we might want to do on the C++ side? (you get a segfault otherwise)

Reply:
Yes, that would be preferable indeed.

Reply:
OK, done